### PR TITLE
Children ccap start date

### PR DESCRIPTION
**Pull request by enyia21**
#### 🔗 Pivotal Tracker ticket
#186822601

#### ✍️ Description
Added a `children-ccap-start-date`

#### 📷 Design reference
[Figma] (https://www.figma.com/file/1oRFY7p6tdGogBJk7Ugle8/IL-CCAP-Full-Flow-(WIP)?type=design&node-id=1190-34116&mode=design&t=Jm7K8bnchInIdzPh-4)

image

#### ✅ Completion tasks

- [ ] Added relevant tests
- [ ] Meets acceptance criteria
- [ ] Added testing instructions to ticket

### CODE REVIEW

File: src/test/java/org/ilgcc/app/journeys/GccFlowJourneyTest.java [hunk 1]
```diff
@@ -39,13 +39,19 @@ void fullGccFlow() {
     testPage.enter("childDateOfBirthDay", "25");
     testPage.enter("childDateOfBirthYear", "2020");
     testPage.selectRadio("needFinancialAssistanceForChild", "Yes");
-    testPage.clickContinue();
+    testPage.clickButton("Continue");
```
Comment (sree-cfa) [inline src/test/java/org/ilgcc/app/journeys/GccFlowJourneyTest.java:42]: Is this change needed?

Comment (enyia21) [reply src/test/java/org/ilgcc/app/journeys/GccFlowJourneyTest.java:42]: No

**Comment by enyia21** [comment]
Updated code to meet the acceptance criteria of having date validation listed in the [pivotal story](https://www.pivotaltracker.com/story/show/186822601).  Also a test has been added to the journey test to verify that there is validation on this date field of the `children-ccap-start-date` screen.

File: src/main/java/org/ilgcc/app/submission/actions/ValidateChildrenCCAPStartDate.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    if (this.isDateInvalid(ccapStartingDate)) {
+      errorMessages.put(INPUT_NAME, List.of("Please check the date entered. It is not a valid date."));
+    }else{
+      DateTimeFormatter dtf = DateTimeFormat.forPattern("MM/dd/yyyy");
+      DateTime dateCCAPStart = dtf.parseDateTime(ccapStartingDate);
+      DateTime present = DateTime.now();
+      DateTime earliest_supported_date = dtf.parseDateTime(EARLIEST_DATE_SUPPORTED);
+
+      if(this.isDateNotWithinSupportedRange(dateCCAPStart, earliest_supported_date, present)){
+        errorMessages.put(INPUT_NAME, List.of(String.format("Please check the date entered. %s is not a supported start date.", ccapStartingDate)));
```
Comment (sree-cfa) [inline src/main/java/org/ilgcc/app/submission/actions/ValidateChildrenCCAPStartDate.java:42]: We prob want a message property here so that it can be translated

File: src/main/java/org/ilgcc/app/submission/actions/VerifyDate.java [hunk 3]
```diff
@@ -16,4 +17,8 @@ protected boolean isDateInvalid(String date) {
     }
     return false;
   }
+
+ protected boolean isDateNotWithinSupportedRange (DateTime date, DateTime earliest_supported_date, DateTime latest_supported_date){
```
Comment (sree-cfa) [inline src/main/java/org/ilgcc/app/submission/actions/VerifyDate.java:21]: let's make these args camel case for consistency

File: src/main/java/org/ilgcc/app/submission/actions/ValidateChildrenCCAPStartDate.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                    (String) inputData.get("ccapStartDay"),
+                    (String) inputData.get("ccapStartYear"));
+
+    if (this.isDateInvalid(ccapStartingDate)) {
+      errorMessages.put(INPUT_NAME, List.of("Please check the date entered. It is not a valid date."));
+    }else{
+      DateTimeFormatter dtf = DateTimeFormat.forPattern("MM/dd/yyyy");
+      DateTime dateCCAPStart = dtf.parseDateTime(ccapStartingDate);
+      DateTime present = DateTime.now();
+      DateTime earliest_supported_date = dtf.parseDateTime(EARLIEST_DATE_SUPPORTED);
```
Comment (sree-cfa) [inline src/main/java/org/ilgcc/app/submission/actions/ValidateChildrenCCAPStartDate.java:39]: I think this and `DateTimeFormatter dtf = DateTimeFormat.forPattern("MM/dd/yyyy");` can also be global `private static final` variables

File: src/main/resources/messages.properties [hunk 5]
```diff
@@ -168,6 +168,8 @@ children-ccap-info.race-ethnicity-help-text=Select all that apply. This is optio
 children-ccap-in-care.title=CCAP in care
 children-ccap-in-care.header=Is {0} already in child care with your chosen provider?
 #children-ccap-start-date
+children-ccap-start-date.title=CCAP Start Date
+children-ccap-start-date.header=When did {0} start care at your chosen provider?
```
Comment (sree-cfa) [inline src/main/resources/messages.properties:172]: looks like the copy is slightly different depending on the last screens input - just checking if that'll be handled in a follow-up ticket

File: src/main/resources/templates/fragments/gcc-icons.html [hunk 6]
```diff
@@ -180,6 +180,12 @@ <h1>Icons</h1>
         </svg>
       </td>
     </tr>
+    <tr>
+      <td></td>
+      <td>
+
+      </td>
+    </tr>
```
Comment (sree-cfa) [inline src/main/resources/templates/fragments/gcc-icons.html:188]: I think there is some content missing here

File: src/main/java/org/ilgcc/app/submission/actions/ValidateChildrenCCAPStartDate.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import org.joda.time.format.DateTimeFormatter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.MessageSource;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+public class ValidateChildrenCCAPStartDate extends VerifyDate {
+@Autowired
```
Comment (sree-cfa) [inline src/main/java/org/ilgcc/app/submission/actions/ValidateChildrenCCAPStartDate.java:24]:
```suggestion
  @Autowired
```

File: src/main/resources/templates/fragments/gcc-icons.html (hunk 6 above)
Comment (sree-cfa) [inline src/main/resources/templates/fragments/gcc-icons.html:188]:
```suggestion
```

File: src/test/java/org/ilgcc/app/journeys/GccFlowJourneyTest.java [hunk 8]
```diff
@@ -39,13 +39,29 @@ void fullGccFlow() {
...
```
Comment (sree-cfa) [inline src/test/java/org/ilgcc/app/journeys/GccFlowJourneyTest.java:42]:
```suggestion
    testPage.clickContinue();
```

**Review by sree-cfa — APPROVED** [review]
